common/thorhelper/thorsoapcall.cpp

```diff
@@ -507,6 +509,32 @@ class BlackLister : public CInterface, implements IThreadFactory
     }
 } *blacklist;
 
+class PersistentHandlerWrapper
+{
+public:
+    PersistentHandlerWrapper(IPersistentHandler* handler) : persistentHandler(handler) { }
+    Owned<IPersistentHandler> persistentHandler;
+};
+
+static Singleton<PersistentHandlerWrapper> persistentHandlerSingleton;
+
+PersistentHandlerWrapper & queryPersistentHandler()
+{
+        return *persistentHandlerSingleton.query([] {
+#ifndef _CONTAINERIZED
+            const IProperties &conf = queryEnvironmentConf();
+            int maxPersistentRequests = conf.getPropInt("maxPersistentRequests", DEFAULT_MAX_PERSISTENT_REQUESTS);
+#else
+            IPropertyTree& conf = queryComponentConfig();
+            int maxPersistentRequests = conf.getPropInt("@maxPersistentRequests", DEFAULT_MAX_PERSISTENT_REQUESTS);
+#endif
+            if (maxPersistentRequests == 0)
+                return new PersistentHandlerWrapper(nullptr);
+            else
+                return new PersistentHandlerWrapper(createPersistentHandler(nullptr, DEFAULT_MAX_PERSISTENT_IDLE_TIME, maxPersistentRequests, PersistentLogLevel::PLogMin, true));
```

Comment:
I don't think the Singleton is a good fit here actually, I may not have realized this code only conditionally required the object, so I can see why you've added the wrapper to workaround the requirements of the Singleton.

As queryPersistentHandler() is not called at a high frequency (only once per CWSCAsyncFor ctor), then a crit + an initialized boolean is prob. best. Can't see the previous commit now, but probably how you had it.